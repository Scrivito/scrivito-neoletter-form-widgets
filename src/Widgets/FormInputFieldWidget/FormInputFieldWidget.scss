.scrivito-neoletter-form-widgets {

	.form-input-container {
		width: 100%;
		position: relative;
		.input-label {
			p {
				display: inline;
			}
		}

		&.floating-label {
			min-height: 65px;
			padding-top: 1.25rem;

			.input-label {
				position: absolute;
				top: 26px;
				left: 0.75rem;
				transition: all 0.2s ease;
				pointer-events: none;
			}

			&.is-selected .input-label {
				top: -7px;
				left: 0;

				.mandatory-container span {
					color: var(--bs-danger, #f03a47);
				}

				.helptext-container {
					display: inline-block;
				}
			}

			.helptext-container {
				pointer-events: auto;
				display: none;
			}

			.mandatory-container {
				pointer-events: auto;

				span {
					color: #333;
				}
			}
		}
	}
}