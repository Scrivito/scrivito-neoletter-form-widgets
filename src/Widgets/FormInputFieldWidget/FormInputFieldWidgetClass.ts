import * as Scrivito from "scrivito";

export const FormInputFieldWidget = Scrivito.provideWidgetClass(
  "FormInputFieldWidget",
  {
    attributes: {
      label: "string",
      placeholder: "string",
      required: "boolean",
      helpText: "html",
      type: [
        "enum",
        {
          values: [
            "email",
            "name",
            "given_name",
            "family_name",
            "middle_name",
            "company",
            "phone_number",
            "custom",
          ],
        },
      ],
      customType: ["enum", { values: ["single_line", "multi_line"] }],
      customFieldName: "string",
    },
  }
);
