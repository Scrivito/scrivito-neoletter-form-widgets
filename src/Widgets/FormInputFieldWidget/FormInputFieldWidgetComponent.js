import * as React from "react";
import * as Scrivito from "scrivito";
import { getFieldName } from "../FormContainerWidget/utils/getFieldName";
import { isCustomType } from "../FormContainerWidget/utils/isCustomType";
import { Mandatory } from "../FormContainerWidget/components/MandatoryComponent";
import { HelpText } from "../FormContainerWidget/components/HelpTextComponent";

Scrivito.provideComponent("FormInputFieldWidget", ({ widget }) => {
  const id = `form_text_input_widget_${widget.id()}`;

  const fieldName = getFieldName(widget);
  const labelOptions = {};
  if (!Scrivito.isInPlaceEditingActive()) {
    labelOptions.htmlFor = id;
  }

  return (
    <div className="mb-3">
      <Scrivito.ContentTag
        content={widget}
        attribute="label"
        tag="label"
        className="form-label"
        {...labelOptions}
      />
      {widget.get("required") && <Mandatory />}

      {widget.get("helpText") && <HelpText widget={widget} />}

      {isCustomType(widget) && widget.get("customType") === "multi_line" ? (
        <textarea
          className="form-control"
          id={id}
          rows="3"
          name={fieldName}
          placeholder={widget.get("placeholder")}
          required={widget.get("required")}
        />
      ) : (
        <input
          className="form-control"
          id={id}
          name={fieldName}
          maxLength={calculateMaxLength(fieldName)}
          placeholder={widget.get("placeholder")}
          type={calculateType(fieldName)}
          required={widget.get("required")}
        />
      )}
    </div>
  );
});

function calculateMaxLength(fieldName) {
  switch (fieldName) {
    case "phone_number":
      return "50";
    case "email":
      return "120";
    default:
      return "250";
  }
}

function calculateType(type) {
  if (type === "email") {
    return "email";
  }

  if (type === "phone_number") {
    return "tel";
  }

  return "text";
}
