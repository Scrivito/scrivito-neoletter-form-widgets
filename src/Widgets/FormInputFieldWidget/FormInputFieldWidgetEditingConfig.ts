import * as Scrivito from "scrivito";
import formInputFieldWidgetIcon from "../../assets/images/form_input_field_widget.svg";
import { isCustomType } from "../FormStepContainerWidget/utils/isCustomType";
import { customFieldNameValidation } from "../FormStepContainerWidget/utils/validations/customFieldNameValidation";
import { typeValidation } from "../FormStepContainerWidget/utils/validations/typeValidation";
import { insideFormContainerValidation } from "../FormStepContainerWidget/utils/validations/insideFormContainerValidation";
Scrivito.provideEditingConfig("FormInputFieldWidget", {
  title: "Neoletter Form Input Field",
  thumbnail: formInputFieldWidgetIcon,
  attributes: {
    required: { title: "Mandatory" },
    type: {
      title: "Input type",
      values: [
        { value: "name", title: "Name" },
        { value: "given_name", title: "Given name" },
        { value: "middle_name", title: "Middle name" },
        { value: "family_name", title: "Family name" },
        { value: "email", title: "Email" },
        { value: "phone_number", title: "Phone number" },
        { value: "company", title: "Company" },
        { value: "custom", title: "Custom" }
      ]
    },
    customType: {
      title: "Custom input type",
      values: [
        { value: "single_line", title: "Single-line" },
        { value: "multi_line", title: "Multi-line" }
      ]
    },
    customFieldName: { title: "Field name" },
    helpText: { title: "Help text" }
  },
  initialContent: {
    label: "Custom field",
    placeholder: "Your custom field",
    type: "custom",
    customType: "single_line",
    customFieldName: "custom_field_name"
  },
  properties: widget => {
    if (widget.get("type") === "custom") {
      return [
        "type",
        "customType",
        "customFieldName",
        "label",
        "placeholder",
        "required",
        "helpText"
      ];
    }

    return ["type", "label", "placeholder", "required", "helpText"];
  },
  validations: [
    insideFormContainerValidation,
    typeValidation,
    [
      "customType",
      (customType, { widget }) => {
        if (isCustomType(widget) && !customType) {
          return "Select the custom input type.";
        }
      }
    ],
    customFieldNameValidation
  ]
});
