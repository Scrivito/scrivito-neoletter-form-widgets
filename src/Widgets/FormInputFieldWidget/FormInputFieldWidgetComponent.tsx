import * as React from "react";
import * as Scrivito from "scrivito";
import { getFieldName } from "../FormStepContainerWidget/utils/getFieldName";
import { isCustomType } from "../FormStepContainerWidget/utils/isCustomType";
import { Mandatory } from "../FormStepContainerWidget/components/MandatoryComponent";
import { HelpText } from "../FormStepContainerWidget/components/HelpTextComponent";
import { FormInputFieldWidget } from "./FormInputFieldWidgetClass";
import { isEmpty } from "../FormStepContainerWidget/utils/lodashPolyfills";
import "./FormInputFieldWidget.scss";

// eslint-disable-next-line @typescript-eslint/no-explicit-any
Scrivito.provideComponent(FormInputFieldWidget, ({ widget, onInputChange }: any) => {
  const id = `form_text_input_widget_${widget.id()}`;
  const fieldName = getFieldName(widget);
  const useFloatingLabel = widget.get("useFloatingLabel");
  const [isSelected, setIsSelected] = React.useState(false);


  const handleChange = (event: React.ChangeEvent<HTMLTextAreaElement | HTMLInputElement>) => {
    setIsSelected(event.target.value !== "");
    onInputChange && onInputChange(getFieldName(widget), event.target.value);
  };
  return (
    <div className={`mb-3 form-input-container ${useFloatingLabel ? 'floating-label' : ''} ${isSelected ? "is-selected" : ""}`} >
      {!isEmpty(widget.get("label")) && (
        <>
          {widget.get("label") &&
            <label htmlFor={id} className="input-label">
              <Scrivito.ContentTag
                attribute="label"
                content={widget}
                tag="span"
              />
              {widget.get("required") && <Mandatory />}
              {widget.get("helpText") && <HelpText widget={widget} />}
            </label>
          }
        </>
      )
      }

      {isCustomType(widget) && widget.get("customType") === "multi_line" ? (
        <textarea
          className="form-control"
          id={id}
          rows={3}
          name={fieldName}
          placeholder={widget.get("placeholder")}
          required={widget.get("required")}
          onChange={handleChange}
        />
      ) : (
        <input
          className="form-control"
          id={id}
          name={fieldName}
          maxLength={calculateMaxLength(fieldName)}
          placeholder={widget.get("placeholder")}
          type={calculateType(fieldName)}
          required={widget.get("required")}
          onChange={handleChange}
        />
      )}
    </div>
  );
});

function calculateMaxLength(fieldName: string): number {
  switch (fieldName) {
    case "phone_number":
      return 50;
    case "email":
      return 120;
    default:
      return 250;
  }
}

function calculateType(type: string): string {
  if (type === "email") {
    return "email";
  }

  if (type === "phone_number") {
    return "tel";
  }

  return "text";
}
