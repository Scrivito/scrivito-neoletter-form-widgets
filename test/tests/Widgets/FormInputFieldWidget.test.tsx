import { screen } from "@testing-library/react";
import * as Scrivito from "scrivito";
import { FormInputFieldWidget } from "../../../src/Widgets/FormInputFieldWidget/FormInputFieldWidgetClass";
import "../../../src/Widgets/FormInputFieldWidget/FormInputFieldWidgetComponent";
import PageRenderer from "../../helpers/pageRenderer";

Scrivito.configure({ tenant: "inMemory" });

const pageRenderer = new PageRenderer();

const widgetProps = {
  label: "Test Label",
  placeholder: "Test placeholder",
  required: true,
  helpText: "Sample help text",
  type: "custom",
  customType: "single_line",
  customFieldName: "custom_test"
};
describe("FormInputFieldWidget Component", () => {
  it("renders input field with custom type & all attributes filled", () => {
    pageRenderer.render({
      body: [new FormInputFieldWidget(widgetProps)]
    });

    const label = screen.getByText(widgetProps.label);
    const mandatoryText = document.querySelector(".text-mandatory");
    const helpTextElement = document.querySelector("i");
    const inputElement = screen.getByPlaceholderText("Test placeholder");

    expect(label).toBeInTheDocument();
    expect(mandatoryText).toBeInTheDocument();
    expect(helpTextElement).toBeInTheDocument();
    expect(inputElement).toBeInTheDocument();
    expect(inputElement).toHaveAttribute("type", "text");
    expect(inputElement).toHaveAttribute("name", "custom_test");
    expect(inputElement).toHaveAttribute("maxLength", "250");
    expect(inputElement).toHaveAttribute("required");
  });

  it("renders input field with email type, not required & no helptext", () => {
    pageRenderer.render({
      body: [
        new FormInputFieldWidget({
          ...widgetProps,
          type: "email",
          required: false,
          helpText: ""
        })
      ]
    });

    const label = screen.getByText(widgetProps.label);
    const mandatoryText = document.querySelector(".text-mandatory");
    const helpTextElement = document.querySelector("i");
    const inputElement = screen.getByPlaceholderText("Test placeholder");

    expect(label).toBeInTheDocument();
    expect(mandatoryText).not.toBeInTheDocument();
    expect(helpTextElement).not.toBeInTheDocument();
    expect(inputElement).toBeInTheDocument();
    expect(inputElement).toHaveAttribute("type", "email");
    expect(inputElement).toHaveAttribute("maxLength", "120");
    expect(inputElement).not.toHaveAttribute("required");
  });

  it("renders input field with telephone type, required & with help text", () => {
    pageRenderer.render({
      body: [
        new FormInputFieldWidget({
          ...widgetProps,
          type: "phone_number",
          required: true,
          placeholder: "Another Test placeholder"
        })
      ]
    });
    const label = screen.getByText(widgetProps.label);
    const mandatoryText = document.querySelector(".text-mandatory");
    const helpTextElement = document.querySelector("i");
    const inputElement = screen.getByPlaceholderText(
      "Another Test placeholder"
    );

    expect(label).toBeInTheDocument();
    expect(mandatoryText).toBeInTheDocument();
    expect(helpTextElement).toBeInTheDocument();
    expect(inputElement).toBeInTheDocument();
    expect(inputElement).toHaveAttribute("type", "tel");
    expect(inputElement).toHaveAttribute("maxLength", "50");
    expect(inputElement).toHaveAttribute("required");
  });

  it("renders correctly", () => {
    const tree = pageRenderer.getAsJSON({
      body: [new FormInputFieldWidget(widgetProps)]
    });
    expect(tree).toMatchSnapshot();
  });
});
