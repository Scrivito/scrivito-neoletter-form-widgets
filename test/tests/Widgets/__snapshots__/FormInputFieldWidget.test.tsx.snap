// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormInputFieldWidget Component renders correctly 1`] = `
<div>
  <div>
    <div
      className="mb-3"
    >
      <label
        className="form-label"
        htmlFor="form_text_input_widget_staticWidgetID"
      >
        Test Label
      </label>
      <span
        className="text-mandatory"
        onBlur={[Function]}
        onFocus={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
      >
        *
      </span>
      <i
        className="bi bi-question-circle"
        onBlur={[Function]}
        onFocus={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
      />
      <input
        className="form-control"
        id="form_text_input_widget_staticWidgetID"
        maxLength={250}
        name="custom_test"
        placeholder="Test placeholder"
        required={true}
        type="text"
      />
    </div>
  </div>
</div>
`;
